lib/fog/compute/google/requests/reset_windows_password.rb

```diff
@@ -0,0 +1,140 @@
+          key = OpenSSL::PKey::RSA.new(2048)
+          modulus, exponent = get_modulus_exponent_in_base64(key)
+
+          # Get Old Metadata
+          old_metadata = server.metadata
+
+          # Create JSON Object with needed information
+          metadata_entry = get_json_string(user, modulus, exponent, email)
+
+          # Create new metadata object
+          new_metadata = update_windows_keys(old_metadata, metadata_entry)
+
+          # Set metadata on instance
+          server.set_metadata(new_metadata, false)
+
+          # Get encrypted password from Serial Port 4 Output
+          # TODO: Should use server.serial_port_output
+          # but unable to get port 4 right now
+
+          # If machine is booting for the first time, there appears to be a
+          # delay before the password appears on the serial port.
+          serial_port_output = get_server_serial_port_output(instance, zone, :port => 4).contents
+          while serial_port_output.empty?
+            sleep(5)
+            serial_port_output = get_server_serial_port_output(instance, zone, :port => 4).contents
+          end
+
+          # Parse and decrypt password
+          enc_password = get_encrypted_password_from_serial_port(serial_port_output, modulus)
+          password = decrypt_password(enc_password, key)
+
+          return password
+        end
+
+        def get_modulus_exponent_in_base64(key)
+          mod = [key.n.to_s(16)].pack("H*").strip
+          exp = [key.e.to_s(16)].pack("H*").strip
+          modulus = Base64.strict_encode64(mod).strip
+          exponent = Base64.strict_encode64(exp).strip
+          return modulus, exponent
+        end
+
+        def get_expiration_time_string
+          utc_now = Time.now.utc
+          expire_time = utc_now + 5 * 60
+          return expire_time.strftime("%Y-%m-%dT%H:%M:%SZ")
+        end
+
+        def get_json_string(user, modulus, exponent, email)
+          expire = get_expiration_time_string
+          data = { 'userName': user,
+                   'modulus': modulus,
+                   'exponent': exponent,
+                   'email': email,
+                   'expireOn': expire }
+          return ::JSON.dump(data)
+        end
+
+        def update_windows_keys(old_metadata, metadata_entry)
+          if old_metadata[:items]
+            new_metadata = Hash[old_metadata[:items].map { |item| [item[:key], item[:value]] }]
+          else
+            new_metadata = {}
+          end
+          new_metadata["windows-keys"] = metadata_entry
+          return new_metadata
+        end
+
+        def get_encrypted_password_from_serial_port(serial_port_output, modulus)
+          output = serial_port_output.split("\n")
+          output.reverse.each do |line|
+            begin
+              entry = ::JSON.parse(line)
+              if modulus == entry["modulus"]
+                return entry["encryptedPassword"]
+              end
+            rescue ::JSON::ParserError
```

Comment:
In general it is somewhat dangerous to use `rescue` as an if statement as you cannot depend on `JSON` to always return `ParserError`. To be clear - I'm not nitpicking here, I just had to learn this the hard way, see: https://bugs.ruby-lang.org/issues/11774

Can we at least scope this down so we don't have to evaluate every line through the parser? E.g. see if string contains the "modulus" and _then_ try to parse it and back off on ParserError?

Reply:
After addressing the above, please add a short log message to the ParserError rescue so it's evident that we have to reparse, e.g. `Parsing modulus out of log string failed, trying to parse next matching line.`

This would help raise any failed assumptions later down the line so we can address them.